runtime/caches.go

```diff
@@ -82,85 +86,123 @@ func (c *connectionCache) Close() error {
-			config["activity"] = activityClient
-			handle, err := c.openAndMigrate(ctx, instanceID, driver, shared, config)
-			if err != nil {
-				return nil, nil, err
-			}
-			conn = &connWithRef{Handle: handle, ref: 0}
-		} else {
-			conn = value.(*connWithRef)
-		}
+	c.lock.Lock()
+	if c.closed {
+		c.lock.Unlock()
+		return nil, nil, errConnectionCacheClosed
+	}
+
+	// check from in-use cache
+	if conn, ok := c.cache[key]; ok {
+		defer c.lock.Unlock()
+		conn.ref++
+		return conn.Handle, c.releaseFn(key, conn), nil
 	}
 
-	// increase reference
-	conn.ref++
-	if found {
+	// check from open cache
+	if value, ok := c.lruCache.Get(key); ok {
+		defer c.lock.Unlock()
+		conn := value.(*connWithRef)
+		conn.ref++
 		// remove from lru-cache
 		c.lruCache.Remove(key)
+		// set in in-use cache
+		c.cache[key] = conn
+		return conn.Handle, c.releaseFn(key, conn), nil
 	}
-	// set in in-use cache
-	c.cache[key] = conn
-	if len(c.cache)+c.lruCache.Len() > c.size {
-		c.logger.Warn("number of in-use connections and open connections exceed total configured size", zap.Int("in-use", len(c.cache)), zap.Int("open", c.lruCache.Len()))
-	}
-	return conn.Handle, func() {
+
+	// connection is not opened
+	// We unlock the mutex before running openAndMigrate and later need it for setting value in cache.
+	// A deadlock is possible if goroutine waiting for the singleflight.Do result also holds the global mutex
+	// so unlock global mutex before entering singleflight.Do even if this means taking lock again for checking cache
+	c.lock.Unlock()
+	dedupe := true // singleflight call was deduped and connection's reference wasn't increased
+	conn, err := c.singleflight.Do(ctx, key, func(ctx context.Context) (*connWithRef, error) {
+		dedupe = false
+		// try cache again
 		c.lock.Lock()
-		defer c.lock.Unlock()
+		conn, ok := c.cache[key]
+		if !ok {
+			if val, ok := c.lruCache.Get(key); ok {
+				conn = val.(*connWithRef)
+			} else {
+				// open a new connection
+				c.lock.Unlock()
+				h, err := c.openAndMigrate(ctx, instanceID, driver, shared, config)
+				if err != nil {
+					return nil, err
+				}
 
-		conn.ref--
-		if conn.ref == 0 { // not in use
-			// add key to lrucache for eviction
-			c.lruCache.Add(key, conn)
-			// delete from in-use cache
-			delete(c.cache, key)
+				conn = &connWithRef{Handle: h, ref: 0}
+				c.lock.Lock()
+			}
+		}
+		conn.ref++
+		// set in in-use cache
+		c.cache[key] = conn
+		if len(c.cache)+c.lruCache.Len() > c.size {
+			c.logger.Warn("number of in-use connections and open connections exceed total configured size", zap.Int("in-use", len(c.cache)), zap.Int("open", c.lruCache.Len()))
 		}
-	}, nil
+		c.lock.Unlock()
+		return conn, nil
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	if dedupe {
+		c.lock.Lock()
+		defer c.lock.Unlock()
+		conn.ref++
+		// remove from lru-cache
+		c.lruCache.Remove(key)
+		// set in in-use cache
+		c.cache[key] = conn
+	}
+	return conn.Handle, c.releaseFn(key, conn), nil
```

Comment:
Is there an unlikely but possible race condition here where the connection is evicted and closed again before the ref is bumped for a deduped case?